Refactor the Calling Java Code from Ballerina guide (#3311)
ballerina-platform/ballerina-dev-website: learn/references/java-interoperability/ballerina-ffi.md

title: Ballerina FFI
description: The reference guide on the list of language features that enable Ballerina developers to call foreign code written in other programming languages.
keywords: ballerina, programming language, ffi, foreign function invocation
permalink: /learn/java-interoperability/ballerina-ffi/
active: ballerina-ffi
redirect_from:
  - /learn/tooling-guide/cli-tools/ballerina-ffi/
  - /learn/tooling-guide/cli-tools/ballerina-ffi
  - /learn/cli-documentation/ballerina-ffi/
  - /learn/cli-documentation/ballerina-ffi
  - /learn/cli-documentation/ballerina-ffi/
  - /learn/cli-documentation/ballerina-ffi
---

Let's look at the list of language features that enable Ballerina developers to call foreign code written in other programming languages. E.g., while the jBallerina compiler allows you to call any `Java` code, the nBallerina compiler will allow you to call any `C` Code.

## The External Function Body
Usually, the body or the implementation of a function is specified in the same source file. The part, which is enclosed by curly braces is called the function body.

```ballerina
function doSomething(int i) returns string {
	...
}
```

> Ballerina also allows you to define a function without a function body and marks it with the `external` keyword to express that the implementation is not provided by the Ballerina source file.

Reviewer: ```suggestion
Ballerina also allows you to define a function without a function body and mark it with the `external` keyword to express that the implementation is not provided by the Ballerina source file.
```

```ballerina
function doSomething(int i) returns string = external;
```

Now, let’s see how you can link this function with a foreign function.

```ballerina
import ballerina/jballerina.java;

function doSomething(int i) returns string = @java:Method {
	name: "doSomethingInJava"
	'class: "a.b.c.Foo"
} external;
```

> The `@java:Method` annotation instructs the jBallerina compiler to link with the `doSomethingInJava` static method in the Java class `a.b.c.Foo`. There are a set of annotations and other utilities available in the `ballerina/jballerina.java` module to make Java interoperability work.  This guide covers most of them.

Reviewer: ```suggestion
The `@java:Method` annotation instructs the jBallerina compiler to link to the `doSomethingInJava` static method in the `a.b.c.Foo` Java class. There are a set of annotations and other utilities available in the `ballerina/jballerina.java` module to make Java interoperability work that is covered in this guide.```

    'class: "java.util.ArrayDeque",
    paramTypes: ["java.util.Collection"]
} external;
```

#### The `paramTypes` Field
You can use the `paramTypes` field to resolve the exact overloaded method. This field is defined as follows.

```ballerina
# The `Class` type represents a fully-qualified Java class name.
public type Class string;

# The `ArrayType` represents a Java array type. It is used to specify parameter
# types in the `Constructor` and `Method` annotations.
#
# + class - Element class of the array type
# + dimensions - Dimensions of the array type
public type ArrayType record {|
    Class 'class;
    byte dimensions;
|};

# Describes a Java constructor. If the `paramTypes` field is not specified,
# then the parameter types are inferred from the corresponding Ballerina function.
#
# + class - The class in which the constructor exists
# + paramTypes - An optional field, which describes the parameter types of the constructor
public type ConstructorData record {|
    Class 'class;
    (Class | ArrayType)[] paramTypes?;
|};
```

As per the above definition, `paramTypes` field takes an array of Java classes or array types. The following table contains more details.

> Java Type | Description | Example

Author: Checked.

       return java:toString(handleEle) ?: "";

   }
}

function newArrayDeque() returns handle = @java:Constructor {
   'class: "java.util.ArrayDeque"
} external;

function pop(handle receiver) returns handle = @java:Method {
   'class: "java.util.ArrayDeque"
} external;

function push(handle receiver, handle element) = @java:Method {
   'class: "java.util.ArrayDeque"
} external;
```

This class presents a much clearer API compared to the previous API. Here, is a sample usage of this class.

```ballerina
public function main() {
   StringStack stack = new();
   stack.push("Ballerina");
   string element = stack.pop();
}
```

### Calling Overloaded Java Methods
The [Instantiating Java Classes](#instantiating-java-classes) section presents how to deal with overloaded constructors. You need to use the same approach to deal with overloaded Java methods. Let’s try to call the overloaded `append` methods in the `java.lang.StringBuffer` class. Here, is a subset of those methods.

```java
StringBuffer append(boolean b);
StringBuffer append(int i);
StringBuffer append(String str);
StringBuffer append(StringBuffer sb);
StringBuffer append(char[] str);
```

> Here, is the set of Ballerina functions that are linked with the above Java methods. Notice the usage of the `paramTypes` annotation field. You can find more details of this field in the [Instantiating Java Classes](#instantiating-java-classes) section.

Author: It's not really a note. The intention is to let the user understand how the `paramTypes` are used.

   name: "append",
   paramTypes: ["java.lang.String"],
   'class: "java.lang.StringBuffer"
} external;

function appendStringBuffer(handle sbObj, handle sb) returns handle = @java:Method {
   name: "append",
   paramTypes: ["java.lang.StringBuffer"],
   'class: "java.lang.StringBuffer"
} external;
```

## Java Exceptions as Ballerina Errors
A function call in Ballerina may complete abruptly by returning an error or by raising a panic. Panics are rare in Ballerina. The best practice is to handle errors in your normal control flow. Raising a panic is similar to throwing a Java exception. The `trap` action will stop a panic and give you the control back in Ballerina and the `try-catch` statement does the same in Java.

Errors in Ballerina belong to the built-int type `error`. The error type can be considered as a distinct type from all other types. The `error` type does not belong to the `any` type, which is the supertype of all other Ballerina types. Therefore, errors are explicit in Ballerina programs and it is almost impossible to ignore them. For more details, see [BBEs](https://ballerina.io/learn/by-example/).

> How are Java exceptions mapped to Ballerina errors?

Author: Removed this question.
